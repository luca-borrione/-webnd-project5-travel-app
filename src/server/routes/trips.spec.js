// const { handleError } = require('../utils/error-utils');

jest.mock('../utils/error-utils');

jest.mock(
  'bent',
  () => () =>
    jest
      .fn()
      .mockImplementation((thumbnail) =>
        thumbnail === 'mock-expired-thumbnail' ? Promise.reject() : Promise.resolve()
      )
);

// eslint-disable-next-line no-promise-executor-return
const flushPromises = () => new Promise((resolve) => setImmediate(resolve));

describe('trips', () => {
  let addTripPostRoute;
  let removeTripPostRoute;
  let restoreTripsPostRoute;

  beforeEach(() => {
    jest.resetModules();
    addTripPostRoute = require('./trips').addTripPostRoute; // eslint-disable-line global-require
    removeTripPostRoute = require('./trips').removeTripPostRoute; // eslint-disable-line global-require
    restoreTripsPostRoute = require('./trips').restoreTripsPostRoute; // eslint-disable-line global-require
  });

  const addTrip = (trip) => {
    const mockRequest = { body: { trip } };
    const mockResponse = { json: jest.fn() };
    addTripPostRoute(mockRequest, mockResponse);
    return mockResponse;
  };

  // const getSavedTrips = () => {
  //   const mockRequest = {};
  //   const mockResponse = { json: jest.fn() };
  //   controller.getSavedTrips(mockRequest, mockResponse);
  //   return mockResponse;
  // };

  const removeTrip = (tripId) => {
    const mockRequest = { body: { tripId } };
    const mockResponse = { json: jest.fn() };
    removeTripPostRoute(mockRequest, mockResponse);
    return mockResponse;
  };

  const restoreTrips = (localStorageTrips) => {
    const mockRequest = { body: { localStorageTrips } };
    const mockResponse = {
      status: jest.fn().mockReturnThis(),
      json: jest.fn(),
    };
    restoreTripsPostRoute(mockRequest, mockResponse);
    return mockResponse;
  };

  describe('addTripPostRoute', () => {
    it('should always respond with a success flag and the array of saved trips', () => {
      const mockResponse = addTrip({ id: 1 });
      expect(mockResponse.json).toHaveBeenCalledTimes(1);
      expect(mockResponse.json).toHaveBeenCalledWith({
        success: true,
        results: { data: [{ id: 1 }] },
      });
    });

    it('should store all the posted trips', () => {
      addTrip({ id: 1 });
      const mockResponse = addTrip({ id: 2 });
      expect(mockResponse.json).toHaveBeenCalledTimes(1);
      expect(mockResponse.json).toHaveBeenCalledWith({
        results: { data: [{ id: 1 }, { id: 2 }] },
        success: true,
      });
    });

    it('should remove the previously stored entry if one with the same id is sent again', () => {
      addTrip({ id: 1 });
      addTrip({ id: 2 });
      const mockResponse = addTrip({ id: 1 });
      expect(mockResponse.json).toHaveBeenCalledTimes(1);
      expect(mockResponse.json).toHaveBeenCalledWith({
        results: { data: [{ id: 2 }, { id: 1 }] },
        success: true,
      });
    });
  });

  describe('restoreTripsPostRoute', () => {
    it('should respond with success flag and the array of the remaining saved trips, if it succeeded in removing a trip', () => {
      addTrip({ id: 1 });
      addTrip({ id: 2 });
      const mockResponse = removeTrip(1);
      expect(mockResponse.json).toHaveBeenCalledTimes(1);
      expect(mockResponse.json).toHaveBeenCalledWith({
        success: true,
        results: { data: [{ id: 2 }] },
      });
    });

    it('should respond with success flag and the array of the remaining saved trips, if the id passed is not corresponding to any saved trip', () => {
      addTrip({ id: 1 });
      addTrip({ id: 2 });
      const mockResponse = removeTrip(3);
      expect(mockResponse.json).toHaveBeenCalledTimes(1);
      expect(mockResponse.json).toHaveBeenCalledWith({
        success: true,
        results: { data: [{ id: 1 }, { id: 2 }] },
      });
    });

    it('should respond with success flag and an empty array as remaining saved trips, if there was no saved trips', () => {
      const mockResponse = removeTrip(3);
      expect(mockResponse.json).toHaveBeenCalledTimes(1);
      expect(mockResponse.json).toHaveBeenCalledWith({
        success: true,
        results: { data: [] },
      });
    });
  });

  describe('restoreTripsPostRoute', () => {
    const mockTrip = {
      thumbnail: 'mock-thumbnail',
      locationInfo: {
        city: 'mock-city',
        country: 'mock-country',
      },
    };

    it('should restore the trips saved in the local storage, if there are no trips saved on the server', async () => {
      const mockResponse = restoreTrips([
        {
          ...mockTrip,
          saved: 'on-local-storage',
        },
      ]);
      await flushPromises();
      expect(mockResponse.json).toHaveBeenCalledTimes(1);
      expect(mockResponse.json).toHaveBeenCalledWith({
        results: {
          data: [
            {
              saved: 'on-local-storage',
              locationInfo: { city: 'mock-city', country: 'mock-country' },
              thumbnail: 'mock-thumbnail',
            },
          ],
        },
        success: true,
      });
    });

    it('should restore the trips saved on the server if present, regardless of the trips saved on the local storage', async () => {
      addTrip({
        ...mockTrip,
        saved: 'on-server',
      });
      const mockResponse = restoreTrips([
        {
          ...mockTrip,
          saved: 'on-local-storage',
        },
      ]);
      await flushPromises();
      expect(mockResponse.json).toHaveBeenCalledTimes(1);
      expect(mockResponse.json).toHaveBeenCalledWith({
        results: {
          data: [
            {
              saved: 'on-server',
              locationInfo: { city: 'mock-city', country: 'mock-country' },
              thumbnail: 'mock-thumbnail',
            },
          ],
        },
        success: true,
      });
    });

    it('should default to an empty array if no trips are saved on the server or on the local storage', async () => {
      const mockResponse = restoreTrips();
      await flushPromises();
      expect(mockResponse.json).toHaveBeenCalledTimes(1);
      expect(mockResponse.json).toHaveBeenCalledWith({
        results: {
          data: [],
        },
        success: true,
      });
    });

    it('should update the thumbnail if the saved link expired', async () => {
      jest.mock('./thumbnail', () => ({
        getThumbnail: jest.fn().mockResolvedValue('mock-updated-thumbnail'),
      }));
      jest.resetModules();
      restoreTripsPostRoute = require('./trips').restoreTripsPostRoute; // eslint-disable-line global-require

      const mockResponse = restoreTrips([
        {
          ...mockTrip,
          thumbnail: 'mock-expired-thumbnail',
        },
      ]);
      await flushPromises();
      expect(mockResponse.json).toHaveBeenCalledTimes(1);
      expect(mockResponse.json).toHaveBeenCalledWith({
        results: {
          data: [
            {
              locationInfo: { city: 'mock-city', country: 'mock-country' },
              thumbnail: 'mock-updated-thumbnail',
            },
          ],
        },
        success: true,
      });
    });

    it('should try to update the thumbnail, if the thumbnail is undefind', async () => {
      jest.mock('./thumbnail', () => ({
        getThumbnail: jest.fn().mockResolvedValue('mock-updated-thumbnail'),
      }));
      jest.resetModules();
      restoreTripsPostRoute = require('./trips').restoreTripsPostRoute; // eslint-disable-line global-require

      const mockResponse = restoreTrips([
        {
          ...mockTrip,
          thumbnail: undefined,
        },
      ]);
      await flushPromises();
      expect(mockResponse.json).toHaveBeenCalledTimes(1);
      expect(mockResponse.json).toHaveBeenCalledWith({
        results: {
          data: [
            {
              locationInfo: { city: 'mock-city', country: 'mock-country' },
              thumbnail: 'mock-updated-thumbnail',
            },
          ],
        },
        success: true,
      });
    });

    it('should override the thumbnail with undefined if the saved link expired and the getThumbnail api rejects', async () => {
      jest.mock('./thumbnail', () => ({
        getThumbnail: jest.fn().mockRejectedValue(),
      }));
      jest.resetModules();
      restoreTripsPostRoute = require('./trips').restoreTripsPostRoute; // eslint-disable-line global-require

      const mockResponse = restoreTrips([
        {
          ...mockTrip,
          thumbnail: 'mock-expired-thumbnail',
        },
      ]);
      await flushPromises();
      expect(mockResponse.json).toHaveBeenCalledTimes(1);
      expect(mockResponse.json).toHaveBeenCalledWith({
        results: {
          data: [
            {
              locationInfo: { city: 'mock-city', country: 'mock-country' },
              thumbnail: undefined,
            },
          ],
        },
        success: true,
      });
    });

    it('a should update the thumbnail if the saved link expired', async () => {
      const { handleError } = require('../utils/error-utils'); // eslint-disable-line global-require
      const mockError = new Error('mock-error-message');

      const localStorageTrips = [mockTrip];
      const mockRequest = { body: { localStorageTrips } };
      const mockStatusJson = jest.fn();
      const mockResponse = {
        status: jest.fn().mockImplementation(() => ({
          json: mockStatusJson,
        })),
        json: jest.fn().mockImplementationOnce(() => {
          throw mockError;
        }),
      };
      restoreTripsPostRoute(mockRequest, mockResponse);

      await flushPromises();

      expect(handleError).toHaveBeenCalledTimes(1);
      expect(handleError).toHaveBeenCalledWith(mockError);
      expect(mockResponse.status).toHaveBeenCalledTimes(1);
      expect(mockResponse.status).toBeCalledWith(500);
      expect(mockStatusJson).toHaveBeenCalledTimes(1);
      expect(mockStatusJson).toHaveBeenCalledWith({
        message: 'mock-error-message',
        success: false,
      });
    });
  });
});
