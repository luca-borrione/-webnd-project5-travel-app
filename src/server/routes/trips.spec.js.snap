// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trips restoreTripsPostRoute should add fresh weather forecast for the departure only, if the return date expired 1`] = `
Object {
  "results": Object {
    "data": Array [
      Object {
        "departureInfo": Object {
          "dateString": "mock-departure-date",
          "weather": "mock-departure-weather",
        },
        "locationInfo": Object {
          "city": "mock-city",
          "country": "mock-country",
        },
        "returnInfo": Object {
          "dateString": "mock-expired-return-date",
          "weather": undefined,
        },
        "thumbnail": "mock-thumbnail",
      },
    ],
  },
  "success": true,
}
`;

exports[`trips restoreTripsPostRoute should add fresh weather forecasts, if both the departure date and the return date are not expired 1`] = `
Object {
  "results": Object {
    "data": Array [
      Object {
        "departureInfo": Object {
          "dateString": "mock-departure-date",
          "weather": "mock-departure-weather",
        },
        "locationInfo": Object {
          "city": "mock-city",
          "country": "mock-country",
        },
        "returnInfo": Object {
          "dateString": "mock-return-date",
          "weather": "mock-return-weather",
        },
        "thumbnail": "mock-thumbnail",
      },
    ],
  },
  "success": true,
}
`;

exports[`trips restoreTripsPostRoute should be able to restore multiple trips 1`] = `
Object {
  "message": "Cannot read properties of undefined (reading 'departureWeather')",
  "success": false,
}
`;

exports[`trips restoreTripsPostRoute should default to an empty array if no trips are saved on the server or on the local storage 1`] = `
Object {
  "results": Object {
    "data": Array [],
  },
  "success": true,
}
`;

exports[`trips restoreTripsPostRoute should not add any weather forecast, if both the departure date and the return date expired 1`] = `
Object {
  "results": Object {
    "data": Array [
      Object {
        "departureInfo": Object {
          "dateString": "mock-expired-departure-date",
          "weather": undefined,
        },
        "locationInfo": Object {
          "city": "mock-city",
          "country": "mock-country",
        },
        "returnInfo": Object {
          "dateString": "mock-expired-return-date",
          "weather": undefined,
        },
        "thumbnail": "mock-thumbnail",
      },
    ],
  },
  "success": true,
}
`;

exports[`trips restoreTripsPostRoute should not add any weather forecast, if the getWeatherForecast api rejects 1`] = `
Object {
  "results": Object {
    "data": Array [
      Object {
        "departureInfo": Object {
          "dateString": "mock-departure-date",
          "weather": undefined,
        },
        "locationInfo": Object {
          "city": "mock-city",
          "country": "mock-country",
        },
        "returnInfo": Object {
          "dateString": "mock-return-date",
          "weather": undefined,
        },
        "thumbnail": "mock-thumbnail",
      },
    ],
  },
  "success": true,
}
`;

exports[`trips restoreTripsPostRoute should override the thumbnail with undefined if the saved link expired and the getThumbnail api rejects 1`] = `
Object {
  "results": Object {
    "data": Array [
      Object {
        "departureInfo": Object {
          "dateString": "mock-departure-date",
          "weather": "mock-departure-weather",
        },
        "locationInfo": Object {
          "city": "mock-city",
          "country": "mock-country",
        },
        "returnInfo": Object {
          "dateString": "mock-return-date",
          "weather": "mock-return-weather",
        },
        "thumbnail": undefined,
      },
    ],
  },
  "success": true,
}
`;

exports[`trips restoreTripsPostRoute should restore the trips saved in the local storage, if there are no trips saved on the server 1`] = `
Object {
  "results": Object {
    "data": Array [
      Object {
        "departureInfo": Object {
          "dateString": "mock-departure-date",
          "weather": "mock-departure-weather",
        },
        "locationInfo": Object {
          "city": "mock-city",
          "country": "mock-country",
        },
        "returnInfo": Object {
          "dateString": "mock-return-date",
          "weather": "mock-return-weather",
        },
        "saved": "on-local-storage",
        "thumbnail": "mock-thumbnail",
      },
    ],
  },
  "success": true,
}
`;

exports[`trips restoreTripsPostRoute should restore the trips saved on the server if present, regardless of the trips saved on the local storage 1`] = `
Object {
  "results": Object {
    "data": Array [
      Object {
        "departureInfo": Object {
          "dateString": "mock-departure-date",
          "weather": "mock-departure-weather",
        },
        "locationInfo": Object {
          "city": "mock-city",
          "country": "mock-country",
        },
        "returnInfo": Object {
          "dateString": "mock-return-date",
          "weather": "mock-return-weather",
        },
        "saved": "on-server",
        "thumbnail": "mock-thumbnail",
      },
    ],
  },
  "success": true,
}
`;

exports[`trips restoreTripsPostRoute should update the thumbnail if the saved link expired 1`] = `
Object {
  "results": Object {
    "data": Array [
      Object {
        "departureInfo": Object {
          "dateString": "mock-departure-date",
          "weather": "mock-departure-weather",
        },
        "locationInfo": Object {
          "city": "mock-city",
          "country": "mock-country",
        },
        "returnInfo": Object {
          "dateString": "mock-return-date",
          "weather": "mock-return-weather",
        },
        "thumbnail": "mock-updated-thumbnail",
      },
    ],
  },
  "success": true,
}
`;

exports[`trips restoreTripsPostRoute should update the thumbnail, if the thumbnail is undefined 1`] = `
Object {
  "results": Object {
    "data": Array [
      Object {
        "departureInfo": Object {
          "dateString": "mock-departure-date",
          "weather": "mock-departure-weather",
        },
        "locationInfo": Object {
          "city": "mock-city",
          "country": "mock-country",
        },
        "returnInfo": Object {
          "dateString": "mock-return-date",
          "weather": "mock-return-weather",
        },
        "thumbnail": "mock-updated-thumbnail",
      },
    ],
  },
  "success": true,
}
`;
